keras_cv/layers/preprocessing/posterization.py

```diff
@@ -0,0 +1,101 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+from tensorflow.keras.__internal__.layers import BaseImageAugmentationLayer
+
+from keras_cv.utils.preprocessing import transform_value_range
+
+
+class Posterization(BaseImageAugmentationLayer):
+    """Reduces the number of bits for each color channel.
+
+    References:
+    - [AutoAugment: Learning Augmentation Policies from Data](
+        https://arxiv.org/abs/1805.09501
+    )
+    - [RandAugment: Practical automated data augmentation with a reduced search space](
+        https://arxiv.org/abs/1909.13719
+    )
+
+    Args:
+        bits: integer. The number of bits to keep for each channel. Must be a value
+            between 1-8.
+        value_range: a tuple or a list of two elements. The value range of input images.
+            When applying the layer, the image values must be scaled to 0-255, with the
+            help of provided `value_range`.
+
+     Usage:
+    ```python
+    (images, labels), _ = tf.keras.datasets.cifar10.load_data()
+    print(images[0, 0, 0])
+    # [59 62 63]
+    # Note that images are Tensors with values in the range [0, 255] and uint8 dtype
+    posterization = Posterization(bits=4, value_range=[0, 255])
+    images = posterization(images)
+    print(images[0, 0, 0])
+    # [48., 48., 48.]
+    # NOTE: the layer will output values in tf.float32, regardless of input dtype.
+    ```
+
+     Call arguments:
+        inputs: input tensor in two possible formats:
+            1. single 3D (HWC) image or 4D (NHWC) batch of images.
+            2. A dict of tensors where the images are under `"images"` key.
+    """
+
+    def __init__(self, bits: int, value_range, **kwargs):
+        super().__init__(**kwargs)
+
+        if not len(value_range) == 2:
+            raise ValueError(
+                "value_range must be a sequence of two elements. "
+                f"Received: {value_range}"
+            )
+
+        if not (0 < bits < 9):
+            raise ValueError(f"Bits value must be between 1-8. Received bits: {bits}.")
+
+        self._shift = 8 - bits
+        self._value_range = value_range
+
+    def augment_image(self, image, transformation=None):
+        image = transform_value_range(
+            images=image,
+            original_range=self._value_range,
+            target_range=[0, 255],
+        )
+        image = tf.cast(image, tf.uint8)
+
+        image = self._posterize(image)
+
+        return transform_value_range(
+            images=image,
+            original_range=[0, 255],
+            target_range=self._value_range,
```

Comment:
lets cast back to compute_dtype before `transform_value_range`.  That way we are certain it is the correct dtype.  transform_value_range can be skipped if value_range is [0, 255]